### PR TITLE
Cache & interpolation

### PR DESCRIPTION
**Pull request by novartole**
Changes:
- Define DatasetType, Column, Table.
- Algorithm to analyze dataset of column.
- Algorithm to find index bounds of given value.
- Approximation comparison with given precision.
- Tests for ApproxCmp, DatasetType, Column.

### CODE REVIEW

File: src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,446 @@
+#![allow(unused)]
+//
+use std::{cmp::Ordering, ops::Deref, sync::Arc};
+//
+type Float = f64;
+type OwnedSet<T> = Arc<[T]>;
+///
+/// Ordering methods (similar to [Ord]) with given approximation.
+pub trait AppoxOrd<Rhs = Self> {
```
[the remark this answers is not in the document]
Comment (novartole) [reply src/cache.rs]: ApproxOrd trait was gone

File: src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+        )+
+    };
+}
+//
+//
+impl_approx_ord! { f32, f64 }
+//
+//
+#[derive(Clone, PartialEq, Debug)]
+enum DatasetType {
```
[the remark this answers is not in the document]
Comment (novartole) [reply src/cache.rs]: DatesetType enum was gone

File: src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            .unwrap_or_default();
+        let table = Table::new(&dbgid, cols);
+        Ok(Self { dbgid, table })
+    }
+}
+//
+//
+impl Cache<Float> {
+    ///
+    /// Returns approximated values based on given ones.
```
[the remark this answers is not in the document]
Comment (novartole) [reply src/cache.rs:83]: added more info to the methods (panics, return types, etc.)

File: src/cache/column.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                    let other = (rhs * base.powi(pr)).trunc();
+                    this.total_cmp(&other)
+                }
+            }
+        )+
+    };
+}
+//
+//
+impl_approx_ord! { f32, f64 }
```
[the remark this answers is not in the document]
Comment (novartole) [reply src/cache/column.rs]: macros was gone

File: src/cache/column.rs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+///
+/// Analyzed dataset, column of [crate::cache::table::Table].
+#[derive(Clone, Debug)]
+pub(in crate::cache) struct Column<T> {
+    inflections: OwnedSet<usize>,
+    data: OwnedSet<T>,
+    precision: u8,
```
[the remark this answers is not in the document]
Comment (novartole) [reply src/cache/column.rs]: another strategy was chosen to compare values, precision field isn't used anymore and has been removed

File: src/cache.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                OwnedSet::from_iter(iter_over_cols)
+            })
+            .unwrap_or_default();
+        let table = Table::new(&dbgid, cols);
+        Ok(Self { dbgid, table })
+    }
+}
+//
+//
+impl Cache<Float> {
```
Comment (a-givertzman) [inline src/cache.rs:81]: Float - is it required, why f64 is not enough

**Review by a-givertzman — CHANGES_REQUESTED** [review]

File: src/cache.rs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+}
+//
+//
+impl<T: PartialOrd> Cache<T> {
+    ///
+    /// Creates an instance using `reader` as the source of values.
+    ///
+    /// # Panics
+    /// Panic occurs if reader produces a non-comparable value (e. g. _NaN_).
+    pub(crate) fn from_reader_with_precision(
```
[the remark this answers is not in the document]
Comment (novartole) [reply src/cache.rs]: fixed and added dbgid to new: 
```rust
Cache.new(&DbgId, impl AsRef<Path>) -> Self { ... }
```

File: src/cache.rs (hunk 6 above)
Comment (novartole) [reply src/cache.rs]: replaced Float with f64

File: src/cache.rs [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    /// This method panics if at least one of the statements is true:
+    /// - `approx_vals` contains a non-comparable value (e. g. _NaN_),
+    /// - reading file at `path` (see [Cache::new]) failed,
+    /// - data of the file is inconsistent (parsing float error or missed data).
+    pub(crate) fn get(&self, approx_vals: &[Option<f64>]) -> Option<Vec<Vec<f64>>> {
+        self.table
+            .get_or_init(|| self.init())
+            .as_ref()
+            .unwrap_or_else(|err| {
+                panic!(
```
Comment (a-givertzman) [inline src/cache.rs:128]: May be `log::error` and return`None` for now ?

**Review by a-givertzman — APPROVED** [review]

File: src/cache.rs (hunk 8 above)
Comment (novartole) [reply src/cache.rs:128]: But how does the caller recognise whether it's out of bounds problem (for instance) or a problem with the file? In runtime the result will be the same (None) and only the log shows the difference

**Pull request merged**